tfe/resource_tfe_team_project_access.go

```diff
@@ -0,0 +1,162 @@
+		ReadContext:   resourceTFETeamProjectAccessRead,
+		UpdateContext: resourceTFETeamProjectAccessUpdate,
+		DeleteContext: resourceTFETeamProjectAccessDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		SchemaVersion: 1,
+
+		Schema: map[string]*schema.Schema{
+			"access": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						string(tfe.TeamProjectAccessAdmin),
+						string(tfe.TeamProjectAccessRead),
+					},
+					false,
+				),
+			},
+
+			"team_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"project_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringMatch(
+					projectIDRegexp,
+					"must be a valid project ID (prj-<RANDOM STRING>)",
+				),
+			},
+		},
+	}
+}
+
+func resourceTFETeamProjectAccessCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	tfeClient := meta.(*tfe.Client)
+
+	// Get the access level
+	access := d.Get("access").(string)
+
+	// Get the project
+	projectID := d.Get("project_id").(string)
+	proj, err := tfeClient.Projects.Read(ctx, projectID)
+	if err != nil {
+		return diag.Errorf(
+			"Error retrieving project %s: %v", projectID, err)
+	}
+
+	// Get the team.
+	teamID := d.Get("team_id").(string)
+	tm, err := tfeClient.Teams.Read(ctx, teamID)
+	if err != nil {
+		return diag.Errorf("Error retrieving team %s: %v", teamID, err)
+	}
+
+	// Create a new options struct.
+	options := tfe.TeamProjectAccessAddOptions{
+		Access:  *tfe.ProjectAccess(tfe.TeamProjectAccessType(access)),
+		Team:    tm,
+		Project: proj,
+	}
+
+	log.Printf("[DEBUG] Give team %s %s access to project: %s", tm.Name, access, proj.Name)
+	tmAccess, err := tfeClient.TeamProjectAccess.Add(ctx, options)
+	if err != nil {
+		return diag.Errorf(
+			"Error giving team %s %s access to project %s: %v", tm.Name, access, proj.Name, err)
+	}
+
+	d.SetId(tmAccess.ID)
+
+	return resourceTFETeamProjectAccessRead(ctx, d, meta)
+}
+
+func resourceTFETeamProjectAccessRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	tfeClient := meta.(*tfe.Client)
+
+	log.Printf("[DEBUG] Read configuration of team access: %s", d.Id())
+	tmAccess, err := tfeClient.TeamProjectAccess.Read(ctx, d.Id())
+	if err != nil {
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Team project access %s no longer exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("Error reading configuration of team project access %s: %v", d.Id(), err)
+	}
+
+	// Update config.
+	d.Set("access", string(tmAccess.Access))
+
+	if tmAccess.Team != nil {
+		d.Set("team_id", tmAccess.Team.ID)
+	} else {
+		d.Set("team_id", "")
+	}
+
+	if tmAccess.Project != nil {
+		d.Set("project_id", tmAccess.Project.ID)
+	} else {
+		d.Set("project_id", "")
+	}
```

Comment:
[Non-blocker] So I'm unfamiliar with this API, but should we not error here? My assumption is that this resource requires to be assigned to at least one team/project (based off the `ForceNew` in the schema)... if that's not the case something was borked and we should let the user know.

Reply:
Er, I suppose setting these attributes to `""` is one way of handling the error since it will recreate the resource.

Reply:
The `project_id` is marked required in the schema and the API contract specifies that there will be an associated project, so the `project_id` should never be nil. Other parts of the codebase seem to follow a similar pattern, where they have the `nil` check to avoid deferencing a nil in the "should never happen" case and relying on the `ForceNew` or other checks to handle it.